lib/internal/util/comparisons.js

```diff
@@ -374,23 +374,48 @@ function setHasEqualElement(set, val1, strict, memo) {
   return false;
 }
 
-// Note: we currently run this multiple times for each loose key!
-// This is done to prevent slowing down the average case.
-function setHasLoosePrim(a, b, val) {
-  const altValues = findLooseMatchingPrimitives(val);
-  if (altValues === undefined)
-    return false;
+// See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#Loose_equality_using
+// Sadly it is not possible to detect corresponding values properly in case the
+// type is a string, number, bigint or boolean. The reason is that those values
+// can match lots of different string values (e.g., 1n == '+00001').
+function findLooseMatchingPrimitives(prim) {
+  switch (typeof prim) {
+    case 'undefined':
+      return null;
+    case 'object': // Only pass in null as object!
+      return undefined;
+    case 'symbol':
+      return false;
+    case 'string':
+      const number = +prim;
+      if (Number.isNaN(number)) {
+        return false;
```

Comment:
☝️ might pluck the `isNaN` reference above.

Reply:
Comment addressed.